### PR TITLE
New Cache Config API

### PR DESCRIPTION
**Pull request by wu-hui**

### CODE REVIEW

File: Firestore/Source/API/FIRLocalCacheSettings+Internal.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+#include <memory>
+#include "Firestore/core/src/api/settings.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FIRPersistentCacheSettings (Internal)
+
+@property(nonatomic, assign)
+    const firebase::firestore::api::PersistentCacheSettings& internalSettings;
```
Comment (ehsannas) [inline Firestore/Source/API/FIRLocalCacheSettings+Internal.h:28]: This is mainly for my benefit (pardon my ignorance. I'm not an obj-c expert).

I was at first a bit confused about having `internalSettings` here as public `@property` and then having `_internalSettings` as a private member in the .mm file.

this results in code like

return **_internalSettings** == otherSettings.**internalSettings**;

IIUC `otherSettings.internalSettings` invokes the getter method, and it's all good. But I wonder if a method is more idiomatic? e.g.

```
@interface FIRPersistentCacheSettings (Internal)
- (const firebase::firestore::api::PersistentCacheSettings&)internalSettings;
@end
```

**Review by ehsannas — COMMENTED** [review]
Looks good... the swift code is somewhat new for me though

File: Firestore/core/test/unit/api/settings_test.cc [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+  EXPECT_EQ(true, settings.persistence_enabled());
+  EXPECT_EQ(copy.cache_size_bytes(), settings.cache_size_bytes());
+  EXPECT_EQ(nullptr, settings.local_cache_settings());
+}
+
+TEST(Settings, CopyAssignmentOperator) {
+  Settings settings;
+  settings.set_host("host");
+  settings.set_ssl_enabled(true);
+  settings.set_local_cache_settings(PersistentCacheSettings{});
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/unit/api/settings_test.cc]: Done.

**Review by wu-hui — COMMENTED** [review]
Thanks!

File: Firestore/core/src/api/settings.h [hunk 3]
```diff
@@ -110,56 +109,50 @@ class LocalCacheSettings {
...
-                         const MemoryCacheSettings& rhs) {
-    return !(lhs == rhs);
-  }
   size_t Hash() const override;
-
- private:
-  friend class Settings;
 };
 
+bool operator!=(const Settings& lhs, const Settings& rhs);
```
Comment (ehsannas) [inline Firestore/core/src/api/settings.h:144]:
```
bool operator==(const Settings& lhs, const Settings& rhs);
```
is missing.

File: Firestore/core/src/api/settings.cc [hunk 4]
```diff
@@ -28,15 +32,159 @@ constexpr bool Settings::DefaultPersistenceEnabled;
...
+        static_cast<const PersistentCacheSettings&>(settings));
+  }
+  UNREACHABLE();
+}
+
 size_t Settings::Hash() const {
+  if (cache_settings_ == nullptr) {
+    return util::Hash(host_, ssl_enabled_, persistence_enabled_,
+                      cache_size_bytes_);
+  }
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply Firestore/core/src/api/settings.cc:76]: I don't know about `util::Hash`, but `std::hash` handles nullptr correctly. ([example](https://godbolt.org/z/478j6xKE5)). I'd unit test it to make sure for `util::Hash`.

File: Firestore/core/src/api/settings.h (hunk 3 above)
Comment (wu-hui) [reply Firestore/core/src/api/settings.h:144]: This is defined in class declaration at line 79.

File: Firestore/core/src/api/settings.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/api/settings.cc]: util::Hash does not handle unique_ptr (unlike std::hash), so we need to deference the pointer. In this case, we would have to dereference a null pointer.

I added a specialization for unique_ptr for this work.

File: Firestore/core/src/util/hashing.h [hunk 5]
```diff
@@ -199,6 +201,12 @@ size_t RankedInvokeHash(K value, HashChoice<5>) {
   return InvokeHash(underlying);
 }
 
+template <typename K>
+auto RankedInvokeHash(const std::unique_ptr<K>& ptr, HashChoice<6>)
+    -> decltype(InvokeHash(*ptr)) {
+  return ptr ? InvokeHash(*ptr) : 23631;
+}
+
```
Comment (ehsannas) [inline Firestore/core/src/util/hashing.h:209]: Will this code ever get executed?

Based on the comments above:

```
 * HashChoice is a recursive type, defined such that HashChoice<0> is the most
 * specific type with HashChoice<1> and beyond being progressively less
 * specific. This causes the compiler to prioritize the overloads with
 * lower-numbered HashChoice types, allowing compilation to succeed even if
 * multiple specializations match.
 ```

And `HachChoice<2>`: `Hashes the given value if it has a specialization of std::hash.`

And `std::hash` does have a specialization for `std::unique_ptr`:
https://en.cppreference.com/w/cpp/memory/unique_ptr/hash

Comment (wu-hui) [reply Firestore/core/src/util/hashing.h:209]: Nope, because all these settings object do not support std::hash, we use our own Hash() functions.

Comment (ehsannas) [reply Firestore/core/src/util/hashing.h:209]: Oh I see... Thanks for the clarification

**Review by ehsannas — APPROVED** [review]

**Pull request merged**